californium-core/src/main/java/org/eclipse/californium/core/network/stack/Block2BlockwiseStatus.java

```diff
@@ -0,0 +1,302 @@
+		int contentFormat = block.getOptions().getContentFormat();
+		int bufferSize = maxBodySize;
+		if (block.getOptions().hasSize2()) {
+			bufferSize = block.getOptions().getSize2();
+		}
+		Block2BlockwiseStatus status = new Block2BlockwiseStatus(contentFormat, bufferSize);
+		status.setFirst(block);
+		Integer observeCount = block.getOptions().getObserve();
+		if (observeCount != null && OptionSet.isValidObserveOption(observeCount)) {
+			// mark this tracker with the observe no of the block it has been created for
+			status.observe = observeCount;
+			exchange.setNotificationNumber(observeCount);
+		}
+		if (block.getOptions().getETagCount() > 0) {
+			// keep track of ETag included in response
+			status.etag = block.getOptions().getETags().get(0);
+		}
+		return status;
+	}
+
+	/**
+	 * Creates a new tracker for retrieving an arbitrary block of a resource.
+	 * 
+	 * @param exchange The message exchange the transfer is part of.
+	 * @param request The request for retrieving the block.
+	 * @param block2 The options for retrieving the block.
+	 * @return The tracker.
+	 */
+	static Block2BlockwiseStatus forRandomAccessRequest(final Exchange exchange, final Request request, final BlockOption block2) {
+		int contentFormat = request.getOptions().getContentFormat();
+		Block2BlockwiseStatus status = new Block2BlockwiseStatus(contentFormat, 0);
+		status.randomAccess = true;
+		status.setCurrentNum(block2.getNum());
+		status.setCurrentSzx(block2.getSzx());
+		return status;
+	}
+
+	private static int determineResponseBlock2Szx(final Exchange exchange, final int preferredBlockSize) {
+		if (exchange.getRequest() != null) {
+			BlockOption block2 = exchange.getRequest().getOptions().getBlock2();
+			if (block2 != null) {
+				LOGGER.log(Level.FINE, "using block2 szx from early negotiation in request: {0}", block2.getSize());
+				return block2.getSzx();
+			}
+		}
+		LOGGER.log(Level.FINE, "using default preferred block size for response: {0}", preferredBlockSize);
+		return BlockOption.size2Szx(preferredBlockSize);
+	}
+
+	/**
+	 * Checks if a given response is a notification that interferes with this
+	 * blockwise transfer.
+	 * 
+	 * @param responseBlock The response block to check.
+	 * @param block2 The block2 option from the response.
+	 * @return {@code true} if the response is a new notification and this transfer's
+	 *         current block number is &gt; 0.
+	 */
+	synchronized boolean isInterferingNotification(final Response responseBlock, final BlockOption block2) {
+		return responseBlock.isNotification() && block2.getNum() == 0 && getCurrentNum() > 0;
+	}
+
+	/**
+	 * Adds the payload of an incoming response to the buffer.
+	 * 
+	 * @param responseBlock The incoming response.
+	 * @param block2 The block2 option contained in the response.
+	 * @return {@code true} if the payload could be added.
+	 */
+	synchronized boolean addBlock(final Response responseBlock, final BlockOption block2) {
```

Comment:
Yes, I was also thinking about the pros and cons of this. The biggest con IMHO is that we cannot be sure that the passed in block2 is really the block2 from the response (which it has to be for the method to produce the correct result).